### PR TITLE
add api-only doc

### PR DESCRIPTION
**Pull request by minrk**
Describe how to use JupyterHub as a "behind the scenes" implementation detail, rather than a user-facing application, like we do for mybinder.org.

Mainly covers different combinations of `JupyterHub.hub_routespec` and `Proxy.extra_routes` config to change:

- disable all Hub UI
- override behavior of "request for a server that's not running"

I think it would be appropriate to discuss permissions and scopes for operating your own UI as an external service here, but not strictly necessary to land a first version of this doc

My goal for this doc: communicate that JupyterHub is designed "API first", and the scope of the UI we ship is basic functionality only, and the intended, supported way to run fancier custom UI is not to customize Hub templates, but rather to run external services that use the Hub API.

### CODE REVIEW

**Comment by minrk** [comment]
The rendered doc: https://jupyterhub--3640.org.readthedocs.build/en/3640/reference/api-only.html

File: docs/source/reference/api-only.md [hunk 1]
```diff
@@ -0,0 +1,124 @@
+(api-only)=
+
+# Deploying JupyterHub in "API only mode"
+
+JupyterHub is a service for deploying and managing Jupyter servers for users.
```
Comment (willingc) [inline docs/source/reference/api-only.md:5]:
```suggestion
As a service for deploying and managing Jupyter servers for users, JupyterHub
```

File: docs/source/reference/api-only.md [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+This functionality is exposed _primarily_ via a [REST API](rest).
```
Comment (willingc) [inline docs/source/reference/api-only.md:6]:
```suggestion
exposes this functionality _primarily_ via a [REST API](rest).
```

File: docs/source/reference/api-only.md [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+JupyterHub also ships with a _basic_ web UI built on that API,
```
Comment (willingc) [inline docs/source/reference/api-only.md:7]:
```suggestion
For convenience, JupyterHub also ships with a _basic_ web UI built using that REST API.
```

File: docs/source/reference/api-only.md [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+enabling users to click a button to quickly start and stop their servers,
```
Comment (willingc) [inline docs/source/reference/api-only.md:8]:
```suggestion
The basic web UI enables users to click a button to quickly start and stop their servers,
```

File: docs/source/reference/api-only.md [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+and admins to perform some basic user and server management tasks.
```
Comment (willingc) [inline docs/source/reference/api-only.md:9]:
```suggestion
and it lets admins perform some basic user and server management tasks.
```

File: docs/source/reference/api-only.md [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+The REST API has always provided functionality not available in the UI,
```
Comment (willingc) [inline docs/source/reference/api-only.md:11]:
```suggestion
The REST API has always provided additional functionality beyond what is available in the basic web UI.
```

File: docs/source/reference/api-only.md [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+and we try to avoid having any functionality only available in the UI and not the API.
```
Comment (willingc) [inline docs/source/reference/api-only.md:12]:
```suggestion
Similarly, we avoid implementing UI functionality that is also not available via the API.
```

File: docs/source/reference/api-only.md [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+With JupyterHub 2.0, this should be fully the case -
```
Comment (willingc) [inline docs/source/reference/api-only.md:13]:
```suggestion
With JupyterHub 2.0, the basic web UI will **always** be composed using the REST API.
```

File: docs/source/reference/api-only.md [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+no UI pages should rely on information not available via the REST API.
```
Comment (willingc) [inline docs/source/reference/api-only.md:14]:
```suggestion
In other words, no UI pages should rely on information not available via the REST API.
```

File: docs/source/reference/api-only.md [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+Previously, some admin UI functionality could only be achieved via admin pages,
+such as paginated requests.
```
Comment (willingc) [inline docs/source/reference/api-only.md:16]: With JupyterHub 2.0, web pages will no longer provide functionality that is unavailable by API.

File: docs/source/reference/api-only.md [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+
+The JupyterHub UI is customizable via extensible HTML [templates](templates),
```
Comment (willingc) [inline docs/source/reference/api-only.md:18]:
```suggestion
## Limited UI customization via templates
The JupyterHub UI is customizable via extensible HTML [templates](templates),
```

File: docs/source/reference/api-only.md [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+but this has some limited scope to what can be customized.
+Adding some content and messages to existing pages is well supported,
+but changing the page flow and what pages are available are beyond the scope of what is customizable.
+
+Increasingly, JupyterHub is used purely as an API for managing Jupyter servers
```
Comment (willingc) [inline docs/source/reference/api-only.md:23]:
```suggestion
## Rich UI customization with REST API based apps
Increasingly, JupyterHub is used purely as an API for managing Jupyter servers
```

**Review by willingc — COMMENTED** [review]
Looks good to me. Made a few suggestions to give users some context. Thanks @minrk.

**Comment by minrk** [comment]
Thanks for the improvements, @willingc!

**Pull request merged**